File: GitToolBox/src/main/java/zielu/intellij/util/ZDateFormatUtil.java
package zielu.intellij.util;

import com.google.common.base.Preconditions;
import com.intellij.util.text.SyncDateFormat;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.Date;
import org.jetbrains.annotations.NotNull;

public class ZDateFormatUtil {
  private static final Duration MINUTE = Duration.ofSeconds(60);
  private static final Duration HOUR = Duration.of(1, ChronoUnit.HOURS);
  private static final Duration HOUR_AND_MINUTE = HOUR.plusSeconds(1);

  private ZDateFormatUtil() {
    //do nothing
  }

  public static String formatPrettyDateTime(@NotNull ZonedDateTime date, @NotNull ZonedDateTime now,
                                            @NotNull SyncDateFormat dateFormat) {
    Preconditions.checkArgument(now.compareTo(date) >= 0, "Date is before now");
    Instant nowInstant = now.toInstant();
    Instant dateInstant = date.toInstant();
    Duration duration = Duration.between(dateInstant, nowInstant);
    if (duration.compareTo(HOUR_AND_MINUTE) < 0) {
      return formatPrettyUntilOneHour(duration);
    } else {
      LocalDate nowDate = nowInstant.atZone(ZoneOffset.UTC).toLocalDate();
      LocalDate referenceDate = dateInstant.atZone(ZoneOffset.UTC).toLocalDate();
      if (nowDate.equals(referenceDate)) {
        return ZResBundle.message("date.format.today");
      } else if (nowDate.minusDays(1).equals(referenceDate)) {
        return ZResBundle.message("date.format.yesterday");
      } else {
        return dateFormat.format(Date.from(dateInstant));
      }
    }
  }

  private static String formatPrettyUntilOneHour(Duration duration) {
    if (duration.compareTo(MINUTE) < 0) {
      return ZResBundle.message("date.format.minutes.ago", 0);
    }
    return ZResBundle.message("date.format.minutes.ago", duration.dividedBy(MINUTE.getSeconds()).getSeconds());
  }
}

File: src/main/kotlin/zielu/intellij/util/ZDateFormatUtil.kt
package zielu.intellij.util

import com.intellij.util.text.SyncDateFormat
import java.time.Duration
import java.time.Instant
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import java.util.Date

internal object ZDateFormatUtil {

  /**
   * Formats relative date-time in following way:
   * 1. under 1 minutes - moments ago
   * 2. 1 minutes to 1 hour - x minutes, one hour ago
   * 3. over 1 hour - today, yesterday, according to date format
   */
  fun formatPrettyDateTime(
    date: ZonedDateTime,
    now: ZonedDateTime,
    dateFormat: SyncDateFormat
  ): String {
    if (now < date) {
      return ZResBundle.na()
    }
    val nowInstant = now.toInstant()
    val dateInstant = date.toInstant()
    val duration = Duration.between(dateInstant, nowInstant)
    return if (duration < HOUR_AND_MINUTE) {
      formatPrettyUntilOneHour(duration)
    } else {
      formatOverOneHour(dateInstant, nowInstant, dateFormat)
    }
  }

  private fun formatPrettyUntilOneHour(duration: Duration): String {
    return if (duration < MINUTE) {
      ZResBundle.message("date.format.minutes.ago", 0)
    } else ZResBundle.message("date.format.minutes.ago", duration.dividedBy(60).seconds)
  }

  private fun formatOverOneHour(date: Instant, now: Instant, dateFormat: SyncDateFormat): String {
    val nowDate = now.atZone(ZoneOffset.UTC).toLocalDate()
    val referenceDate = date.atZone(ZoneOffset.UTC).toLocalDate()
    return when {
      nowDate == referenceDate -> {
        ZResBundle.message("date.format.today")
      }
      nowDate.minusDays(1) == referenceDate -> {
        ZResBundle.message("date.format.yesterday")
      }
      else -> {
        dateFormat.format(Date.from(date))
      }
    }
  }

  fun formatBetweenDates(date1: ZonedDateTime, date2: ZonedDateTime): String {
    val duration = Duration.between(date1, date2)
    return if (duration.isNegative) {
      // TODO: should handle future - in a x minutes
      ZResBundle.na()
    } else {
      formatRelativePastDuration(duration)
    }
  }

  private fun formatRelativePastDuration(duration: Duration): String {
    return when {
      duration < MINUTE -> {
        ZResBundle.message("date.format.moments.ago")
      }
      duration < HOUR -> {
        ZResBundle.message("date.format.n.minutes.ago", duration.dividedBy(MINUTE))
      }
      duration < DAY -> {
        ZResBundle.message("date.format.n.hours.ago", duration.dividedBy(HOUR))
      }
      duration < WEEK -> {
        ZResBundle.message("date.format.n.days.ago", duration.dividedBy(DAY))
      }
      duration < MONTH -> {
        ZResBundle.message("date.format.n.weeks.ago", duration.dividedBy(WEEK))
      }
      duration < YEAR -> {
        ZResBundle.message("date.format.n.months.ago", duration.dividedBy(MONTH))
      }
      else -> {
        ZResBundle.message("date.format.n.years.ago", duration.dividedBy(YEAR))
      }
    }
  }
}

private val MINUTE = Duration.ofSeconds(60)
private val HOUR = Duration.of(1, ChronoUnit.HOURS)
private val HOUR_AND_MINUTE = HOUR.plusSeconds(1)
private val DAY = Duration.of(1, ChronoUnit.DAYS)
private val WEEK = Duration.of(7, ChronoUnit.DAYS)
private val MONTH = Duration.of(30, ChronoUnit.DAYS)
private val YEAR = Duration.of(365, ChronoUnit.DAYS)
